### DISCUSSION TITLE
Implementing User Authentication

### DISCUSSION THREAD

**Discussion by sauron136** [vitalik/django-ninja#740] [Show and tell]
Hi Vitaliy, I just got started with django-ninja this week and so far it's been good and I am getting a hang of it. However, I have a little trouble implementing User Authentication and Authorization. How does it work? I have a User model already and schema. What do I do from here? Meanwhile, this is my repo: https://github.com/KrystianMaccs/cinema.git

**Comment by baseplate-admin** [comment]
Hi thanks for taking an interest in `django-ninja`,

So basically you have to create your own `Token` model instance.

Here's how i did it:
```python

class Token(models.Model):
    token = models.CharField(
        default=partial(
            get_random_string,
            16,
        ),
        unique=True,
        max_length=16,
        editable=False,
    )
    user = models.OneToOneField(CustomUser, on_delete=models.CASCADE, related_name="user")

    def __str__(self) -> str:
        return f"User : {self.user.username} | Token : {self.token}"

    class Meta:
        indexes = [
            models.Index(fields=["token"]),
        ]
        verbose_name = _("token")
        verbose_name_plural = _("tokens")
```

Now you have to setup token claims. 

[`django-ninja` provides you with authentication class](https://django-ninja.rest-framework.com/guides/authentication/). You need to override it.

```python
import logging
from typing import Any

from apps.user.models import CustomUser
from ninja.compatibility import get_headers
from ninja.security import HttpBearer

from django.conf import settings
from django.contrib.auth.models import AnonymousUser
from django.http import HttpRequest

from .models import Token


class AuthBearer(HttpBearer):
    def authenticate(
        self,
        request: HttpRequest,
        token: str,
    ) -> CustomUser | AnonymousUser:
        try:
            token_data = Token.objects.get(token=token)
            return token_data.user

        except Token.DoesNotExist:
            return AnonymousUser
```

Then when you have a protected route you just add Auth class to it :)

```python
from .models import Token
from ninja import Router

from django.http import HttpRequest, HttpResponse

from .auth import AuthBearer

router = Router()


@router.delete("/logout", auth=AuthBearer())
def post_user_logout_info(request: HttpRequest) -> HttpResponse:
    token: Token = Token.objects.get(user=request.auth)
    token.delete()
    return HttpResponse("Successful", status=200)
 ```

If you need a reference you can see my project : 
* [models.py](https://github.com/baseplate-admin/CoreProject/blob/e67800c8a38989f668f71d7549dcdd94b6941eb1/backend/django_core/apps/api/models.py)
* [auth.py](https://github.com/baseplate-admin/CoreProject/blob/e67800c8a38989f668f71d7549dcdd94b6941eb1/backend/django_core/apps/api/auth.py)
* [endpoints (login,logout,signup)](https://github.com/baseplate-admin/CoreProject/tree/e67800c8a38989f668f71d7549dcdd94b6941eb1/backend/django_core/apps/api/views/user)

**Comment by sauron136** [comment]
Ah! Beautiful! Thanks for this. I will implement this at once. Meanwhile do
you min taking a look at my project and letting me know if I have got
it right, especially the apps/movies... I am not so certain if I have got
the routing done correctly. I tried to do it according to the documentation
but it gave me errors. Had to leave it the way it is now.
https://github.com/KrystianMaccs/cinema

Thanks.

**Reply by baseplate-admin** [reply]
Do you have the error stacktrace?

**Comment by sauron136** [comment]
I am not on the system now. But the error said module couldn't be found.
And this was what I did. I created api file in my movies app and then
another api file in the main project next to the urls file. In the
movies.api file, I used Router() and then in the main project I used
NinjaAPI.

**Reply by baseplate-admin** [reply]
Ahha. This is probably due to the django handles importing. I can take a look after 1-2 hour. Currently in work

**Comment by sauron136** [comment]
Alright. That would be great. Thank you very much. 😁

**Reply by baseplate-admin** [reply]
I opened an issue. See if that fixes

**Comment by sauron136** [comment]
Hi, when I tried to implement your authentication logic, and run server, I
get this error:
ImportError: email-validator is not installed, run `pip install
pydantic[email]`

**Reply by baseplate-admin** [reply]
pip jnstall it?

**Comment by sauron136** [comment]
Yeah, done. However, I have another question that is unrelated to
django-ninja. I have 2 databases: postgres and mongodb on this django
project. I want to sync both databases such that once a model instance
is created, saved or updated on postgres, it automatically gets saved
on mongodb. How can i do this?

**Reply by baseplate-admin** [reply]
I have a counter question. Why do you need mongodb?

**Comment by sauron136** [comment]
For easy query. It's a project I was asked to do and I just need ideas.

**Reply by baseplate-admin** [reply]
Setup django signals, so Whenever you save an object it gets created to mongodb.

**Comment by sauron136** [comment]
Yeah, but I gotta link both databases somehow so that when the signal is
fired, the changes take place. I have the signals already, it's just how to
do the linking that eludes me.

**Reply by baseplate-admin** [reply]
Did you register the signal?

**Comment by sauron136** [comment]
Okay. Let me go ahead and register the signals.
Thanks for your help

**Comment by baseplate-admin** [comment]
@KrystianMaccs if you have an answer please mark an answer. Helps other find this later :)

**Comment by sauron136** [comment]
done
